homeassistant/components/system_log.py

```diff
@@ -0,0 +1,127 @@
+    DOMAIN: vol.Schema({
+        vol.Optional(CONF_MAX_ENTRIES,
+                     default=DEFAULT_MAX_ENTRIES): cv.positive_int,
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+SERVICE_CLEAR_SCHEMA = vol.Schema({})
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Set up the logger component."""
+    conf = config.get(DOMAIN)
+
+    if conf is None:
+        conf = CONFIG_SCHEMA({})
+
+    if DATA_SYSTEM_LOG not in hass.data:
+        hass.data[DATA_SYSTEM_LOG] = LogErrorHandler()
+
+    handler = hass.data[DATA_SYSTEM_LOG]
+    handler.resize(conf.get(CONF_MAX_ENTRIES))
+
+    hass.http.register_view(AllErrorsView(handler))
+    yield from hass.components.frontend.async_register_built_in_panel(
+        'system-log', 'system_log', 'mdi:monitor')
+
+    @asyncio.coroutine
+    def async_service_handler(service):
+        """Handle logger services."""
+        # Only one service so far
+        handler.records.clear()
+
+    descriptions = yield from hass.async_add_job(
+        load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'services.yaml'))
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_CLEAR, async_service_handler,
+        descriptions[DOMAIN].get(SERVICE_CLEAR),
+        schema=SERVICE_CLEAR_SCHEMA)
+
+    return True
+
+
+def _figure_out_source(record):
+    # If a stack trace exists, extract filenames from the entire call stack.
+    # The other case is when a regular "log" is made (without an attached
+    # exception). In that case, just use the file where the log was made from.
+    if record.exc_info:
+        stack = [x[0] for x in traceback.extract_tb(record.exc_info[2])]
+    else:
+        stack = [record.pathname]
+
+    # Iterate through the stack call (in reverse) and find the last call from
+    # a file in HA. Try to figure out where error happened.
+    for pathname in reversed(stack):
+
+        # Try to match with a file within HA
+        match = re.match(r'.*/homeassistant/(.*)', pathname)
+        if match:
+            return match.group(1)
+
+    # Ok, we don't know what this is
+    return 'unknown'
+
+
+def _exception_as_string(exc_info):
+    buf = StringIO()
+    if exc_info:
+        traceback.print_exception(*exc_info, file=buf)
+    return buf.getvalue()
+
+
+def _convert(record):
+    return {
+        'timestamp': record.created,
+        'level': record.levelname,
+        'message': record.getMessage(),
+        'exception': _exception_as_string(record.exc_info),
+        'source': _figure_out_source(record),
+        }
+
+
+class AllErrorsView(HomeAssistantView):
+    """Get all logged errors and warnings."""
+
+    url = "/api/error/all"
+    name = "api:error:all"
+
+    def __init__(self, handler):
+        """Initialize a new AllErrorsView."""
+        self.handler = handler
+
+    @asyncio.coroutine
+    def get(self, request):
+        """Get all errors and warnings."""
+        return self.json(list(map(_convert, self.handler.records)))
```

Comment:
I prefer list comprehensions over map.

```python
return self.json([_convert(record) for record in self.handler.records])
```

Reply:
Yes, I'll fix this one as well. I've coded so much scheme/racket in my days so I tend to fall back to these old primitives...